_posts/2022-03-21-getting-started-with-protobufs-for-typescript.md

```diff
@@ -0,0 +1,231 @@
+---
+layout: post
+title: Getting Started with Protobufs and Typescript
+description: Tutorial for how to start using Protobufs with Typescript
+date: 2022-03-21
+author: Ed Newsom
+highlight: monokai
+image: img/blog/protobuf-ts/protobuf-ts.jpg
+tags: ["protobuf", "typescript", "tutorial", "protoc", "npm"]
+---
+
+![Protobuf and TypeScript](/img/blog/protobuf-ts/protobuf-ts.png)
+
+# Getting Started with Protobufs and Typescript
+Getting Protobufs to work with Typescript for the first time is not very straightforward and this article runs through the minimum steps needed to make it work.  Hopefully after reading this and working through the code, you will be confident enough to start using Protobufs in your own projects.
+
+[Click here to skip to the technical bits](#making-it-work) 
+
+or [Click here for the code repository in Github and to skip all the words](https://github.com/MechanicalRock/protobuf-example-ts)
+
+## How Did We Get Here?
+A while ago I was asked to build a specific type of data API that supports Protobufs. My first reaction was to groan because I had tried to learn Protobufs previously and failed twice.  Admittedly on both occasions I did not have an application to focus on and other shiny things distracted me. Before trying again I asked some of my infinitely more experienced colleagues here at [Mechanical Rock](https://mechanicalrock.io) if it was just me or is the documentation for learning Protobufs from zero really '@#^%'? There was some great discussion on this and they sent me on my way with more than a few useful hints.
+
+Having got to the point where I understand enough to use them in anger, I realised that trying to learn Protobufs with Typescript first was the problem because that is where the doc's fall short. I actually had to revert to my first love, Python, before I could transfer that newly minted knowledge to TypeScript.
+
+## Other Stuff & Assumptions
+I am going to skip over the why Protobufs vs XML, the details of the [proto3 data structure](https://developers.google.com/protocol-buffers/docs/proto3) and the like which you can find in plenty of other articles.  I'm assuming you know: node.js, Typescript, npm and Jest. I am using Mac/Linux here (sorry Windows users but you can deal with your own world of pain). I am also shamelessly basing this on the [Google Protobuf Tutorials](https://developers.google.com/protocol-buffers/docs/tutorials) because I am sure you have already been there, hence using the Address Book example.
+
+# Making It Work
+We're going to work through an end to end process to demonstrate the minimum of using Protobufs with TypeScript. Check out the [repository on Github](https://github.com/MechanicalRock/protobuf-example-ts) which brings this all together.
+
+1. [Define the data structure for your messages](#define-the-data-structure-for-your-messages)
```

Comment:
Do all the anchor references work?

I don't see them lower down, and can't remember if they are auto-generated.

Reply:
Yes, checked, all working.